internal/npm_install/npm_install.bzl

```diff
@@ -192,11 +212,21 @@ yarn_install = repository_rule(
             mandatory = True,
             single_file = True,
         ),
-        "prod_only": attr.bool(
-            default = False,
-            doc = "Don't install devDependencies",
-        ),
         "data": attr.label_list(),
+        "workspace_node_modules": attr.bool(
+          doc = """When set to true yarn will run in the directory where the
```

Comment:
this is just "vendoring the dependencies" right?
I wouldn't expect bazel to be involved in running yarn in this case. Up to the user to vendor things into their source directory (I'd put them in `/third_party/npm/node_modules` in a monorepo)
so I'd expect the bazel setup to be only a `filegroup` that points there.

Reply:
Not quite, the main idea is trying to address a suboptimal workflow with how the current `yarn_install` rules work: 

One is, there is no `repository_cache` support, so the dependencies do not get persisted outside of the bazel external repositories, there are cases where all dependencies will have to be reinstalled even if no dependencies have changed. 

The second issue is, that if dependencies have changed then it will currently discard the whole of `node_modules` as well as `yarn_cache`.

The third issue is when using yarn to add/remove a package via the recommended approach `bazel run @nodejs//:bin/yarn -- add package` this is already adding a `node_modules` to the workspace as this command is running within the workspace and not with the external repository. And then when a node_js binary is started that depends on node_modules managed via bazel then bazel has no idea about the `node_modules` that have been added in the workspace by the previous command, nor does it reuse the same cache, so it just goes and reinstalls all dependencies again from scratch.

This leads to a suboptimal experience in a few cases, it makes caching more tricky in ci, it consumes additional bandwith because there is a lot of unnecessarily fetching of packages and the main reason it leads to a really slow workflow when packages are added/removed as in the current setup all dependencies will get reinstalled twice, once when invoking yarn and once on `bazel run :target`

I am also happy to create an issue on that topic first, or even a small design document, as I do realise it is a bigger change, before going ahead with progressing the PR. I just needed to have some improvements now as this was one of the main critique points about the current setup.

Reply:
Yes bazel would point yarn at a cached node_modules so to speak, within the repo, but it is not checked in, it is on gitignore and just lives on each developers machine.

Reply:
maybe https://github.com/bazelbuild/rules_nodejs/commit/3946a16920f6d1383b99995420d2a358b3cd4e8b has already fixed this? instead of yarn using a cache in the project we can just use the standard global yarn cache on the machine

Reply:
Yeah I was considering this as well and maybe it is the best trade-off (at least until yarn Plug'n'Play is released). The commit you linked does indeed improve the situation but it still does not allow for a nice workflow when one uses `bazel run @nodejs//:bin/yarn -- add package` as the downloaded package will still be copied into two places, but at least it just downloads it once.

Reply:
Hmm, the symlink approach does seem to cause smaller other issues, getting some i/o error sometimes, where bazel says that the files are a symlink and a resolved file at the same time. Rerunning fixes this. I will have a think if I can come up with a better strategy that resolves any last workflow issues and not having to do any "hacky" symlinks.